[labs/react] Fix type compatibility with Preact when adding children

augustjk opened the pull request:
Fixes #4138

Content placed in children position in JSX are typed differently depending on whether it's React or Preact project. Children typing in Preact projects were incompatible with `children` prop of our wrapped components as they only accepted `React.ReactNode` which was more restrictive than what Preact considers as valid renderable child.

This PR adds children typing that's compatible with both React and Preact.

~~Testing for this fix working in Preact was done manually in a working branch for https://github.com/lit/lit/pull/4111. I'd be okay with holding off on merging this until I can do the updates to that branch to demonstrate the fix.~~
Mentioned PR has been merged and this branch has been rebased to that.

AndrewJakubowicz commented on packages/labs/react/src/create-component.ts:
```diff
@@ -48,15 +48,25 @@ export type ReactWebComponent<
+  // Omit keyof E to prefer provided event handler mapping over React's
+  // built-in event handler props.
+  | keyof E
+  // Omit children to replace with out own that's compatible with Preact.
```
Typo: `with out own` -> `with our own`

augustjk replied:
Thanks for catching, done!

AndrewJakubowicz commented on packages/labs/react/src/create-component.ts:
```diff
@@ -48,15 +48,25 @@ export type ReactWebComponent<
-// Acceptable props to the React component. Omit keyof E from HTMLAttributes to
-// prefer provided event handler mapping over React's built-in event handler
-// props.
+// Child type compatible with both React and Preact.
```
Is it worth adding some additional details/references for how/where this type was derived from?

augustjk replied:
Done.

AndrewJakubowicz reviewed (approved):
LGTM

It might be nice to land the Preact example, then land this with the children added to the example, but not too fussed with order.

Great stuff!

augustjk commented:
Rebased off of main with preact example merged, and addressed TODO that adds an element child to the wrapped component. Type test for it is passing. https://github.com/lit/lit/actions/runs/6005198623/job/16287421462?pr=4142#step:8:305
